### ISSUE TITLE
Retrieving one row takes almost half minute

### ISSUE THREAD

**Issue by jequiroga** [CartoDB/odbc_fdw#75]
I have in production a Postgres 10 database up to date and odbc_fdw library compiled with your current master branch (@ Feb, 27th, 2018). I used your previous version and since middle of January queries from postgres to SQL Server 2005 are taking at least 300x more time than before just retrieving one row (The query is always returning one row).

### Steps to Reproduce

1. Install CentOs 7 and exclude from CentosRepo-Base postgres
2. Add The postgres official repository for CentOS 7 and install postgres 10
3. Install odbc_fdw but at the end uninstall all dev packages (gcc, postgresql-10-dev, inixODBC-dev, etc.)
4. Configure UNIX-ODBC to reach MSSQL-Server 2005 database
5. Install ODBC Driver for MS-SQL Server (msodbc /freetds) 
6. Install odbc_fdw extension in postgres
7. Create a Foreign server using the odbc_fdw extension, and a query using the foreign server 
8. Create a query that contains some views or tables to SQL-Server

### Current Result
I'm actually using just two queries all of them retrieve just one record. One do it using three seconds, the critical one use 27 secs. In MSSQL-Server those exact queries takes less than a second, in fact in MSSQL Server Manager there is 00:00 secs doing one record and less than a second retrieving all rows.

Have you ever experienced such bad performance during your tests?. I have in development (Windows machine Postgres 9.6, same MSSQL-Server database as production postgres server) using ogr extension and the retrieval took less than a second (from 800ms to 1 sec). Using the same query your extension takes 25 secs. I added the analyze result in production.

"Foreign Scan on ra.clients_csgor_view  (cost=25.00..17918.00 rows=17893 width=378) (actual time=7259.219..24879.050 rows=1 loops=1)"
"  Output: client_id, client_sin, first_name, last_name, last_app_id, app_program_year_id, app_program_year, app_award_limit, app_status_id, current_program_year_id, current_program_year, pd_established"
"  Filter: ((clients_csgor_view.client_sin)::text = '739533537'::text)"
"  Rows Removed by Filter: 17892"
"  Foreign Table Size: 17893"
"Planning time: 201.383 ms"
"Execution time: 25078.335 ms"

This is the result using a foreign server with FreeTDS, not Microsoft MS-MSQL Server ODBC driver for Unix, to eliminate the ODBC Driver variable from the issue.

"Foreign Scan on ra.clients_csgor_view  (cost=25.00..17918.00 rows=17893 width=1346) (actual time=7122.452..24665.566 rows=1 loops=1)"
"  Output: client_id, client_sin, first_name, last_name, last_app_id, app_program_year_id, app_program_year, app_award_limit, app_status_id, current_program_year_id, current_program_year, pd_established"
"  Filter: ((clients_csgor_view.client_sin)::text = '739533537'::text)"
"  Rows Removed by Filter: 17892"
"  Foreign Table Size: 17893"
"Planning time: 217.162 ms"
"Execution time: 24822.832 ms"

The exact same MSSQL-Server database using ogr extension (only for Windows, that is why I had to use yours) in a postgres 9.6 with ogr_fdw.

"Foreign Scan on ra.clients_csgor_view  (cost=25.00..17918.00 rows=17893 width=378) (actual time=51.760..235.947 rows=1 loops=1)"
"  Output: client_id, client_sin, first_name, last_name, last_app_id, app_program_year_id, app_program_year, app_award_limit, app_status_id, current_program_year_id, current_program_year, pd_established"
"  Filter: ((clients_csgor_view.client_sin)::text = '721362009'::text)"
"  Rows Removed by Filter: 17892"
"  Foreign Table Size: 17893"
"Planning time: 409.593 ms"
"Execution time: 491.747 ms"

I Read that using the option "use_remote_estimate" the performance can be improved, but your extension seems that does not allow it; not in the foreign server options. in the table I added, but no effect whatsoever.

### Expected result

To have a performance that takes 1 sec at most (is just one record!)

### System trying to connect to and version
Production Server:
- CentOS 7 minimal (updated up to date)
- Postgresql 10 (updated up to date)
- odbc_fdw (compiled using sourced @ Feb 27th, 2018)
- unix-ODBC (latest available on CentOS repos)
- Microsoft Unix ODBC (msodbcsql-13.1.9.1-1.x86_64.rpm and FreeTDS 7.6)

MSSQL Server 2005 (all service packs applied)
- Windows 7 (Up to date)
- CPU Intel core i5 8GBytes of RAM

Development
- Windows 10 (up to date)
- Postgres 9.6
- ogr_fdw 
- ODBC 64 bits (SQL Server native Client 11.0).

If you need anything else, please let me know. I'm not an expert on Postgres, so I could ask more detailed directions.

**Comment by rafatower** [comment]
@jequiroga thanks a lot for reporting. Could you please try a debug build?

```
$ DEBUG=true make
```

then raise the debug level ('verbosity') of postgres to `DEBUG1`:

```
SET client_min_messages TO DEBUG1; -- IIRC, I think there's no need to tweak the server config.
```

and share the extra information with us? mind that there may be connection information in those traces.

Particularly interested in this trace:
https://github.com/CartoDB/odbc_fdw/blob/e911da73a2f5ed71d6a59e6a7ded1e239e4492ed/odbc_fdw.c#L1448

@pramsey looks like we broke something. Do you have any idea what it could be?

**Comment by pramsey** [comment]
Weird. Our changes were very limited, just enough to get building under the new versions. The only one I'm even a little looking at would affect the quals we push, and we can see in the explain that the qual is in fact rational. Actually, maybe not? If you run the explain with the `verbose` option as well we can see what gets sent to the remote:

`explain (analyze, verbose) select ...`

**Comment by rafatower** [comment]
Oh, you're right! the `explain (analyze, verbose) ...` is a much simpler approach than the `DEBUG` build to get the info we need to debug and compare against.

**Comment by jequiroga** [comment]
@rafatower This is the select that gave the results shown in my first entry. EXPLAIN (ANALYZE, VERBOSE) SELECT * FROM ra.clients_csgor_view WHERE client_sin = '721362009', So I have to recompile with the debugging. I'm installing ogr_fdw, but the dependencies are wildly bigger that was the reason I selected this one, If I have enough time I'll try with that one to compare and see if there is any significant difference.

**Comment by jequiroga** [comment]
Seemed that there is a problem with "SET client_min_messages TO DEBUG1;", in PgAdmin or in psql the connection get lost, and this is all I could get from psql. I tried twice and both terminate the connection abnormally.

EXPLAIN (ANALYZE, VERBOSE) SELECT * FROM ra.clients_csgor_view WHERE client_sin = '721362009';
DEBUG:  odbcGetForeignRelSize
DEBUG:  odbcGetTableOptions
DEBUG:  odbcGetOptions
DEBUG:  extract_odbcFdwOptions
DEBUG:  CONN STR: database=ATBC;DSN=ATWIS_Test;UID=<ms_sql_user>;PWD=<mssql_user_user_passwd>;DSN=ATWIS_Test;UID=<ms_sql_user>;PWD=<mssql_user_user_passwd>
DEBUG:  Successful result: Connecting to driver
DEBUG:  getQuoteChar
DEBUG:  getNameQualifierChar
DEBUG:  Count query: SELECT COUNT(*) FROM (SELECT ac.intAdult_Id AS "client_id",
      ISNULL(ac.chrSIN_No, '') AS "client_sin",
      REPLACE(REPLACE(ISNULL(ac.chvFst_Name, ''), CHAR(146), ''''), CHAR(200), 'È') AS "first_name",
      REPLACE(ISNULL(ac.chvLst_Name, ''), CHAR(146), '''') AS "last_name",
      ISNULL(ap.intRfral_Mstr_Id, 0) AS "last_app_id",
      ISNULL(rm.intCSG_Program_Year_Id, 0) AS "app_program_year_id",
      ISNULL(rm.chvDesc, '') AS "app_program_year",
      rm.mnyAward_Limit AS "app_award_limit",
      rm.intRfral_Status_Id AS "app_status_id",
      cpy.intCSG_Program_Year_Id AS "current_program_year_id",
      cpy.chvDesc AS "current_program_year",
      ISNULL(vcp.isPD_Established, 0) AS "pd_established"
    FROM dbo.tbl_Adult_Client ac
      INNER JOIN tbl_Adult_Program ap ON ap.intAdult_Id = ac.intAdult_Id
        AND ap.intProgram_Id = 2
      INNER JOIN vt_Referral rm ON rm.intRfral_Mstr_Id = ap.intRfral_Mstr_Id
      LEFT JOIN vPD_Client_Program vcp ON vcp.intAdult_Id = ac.intAdult_Id
        AND vcp.intProgram_Id = 2
      LEFT JOIN tbl_CSG_Program_Year cpy ON cpy.bitIs_Default = 1
    WHERE ac.intAdult_Id > 0 AND ISNULL(LTRIM(RTRIM(ac.chrSIN_No)), '') <> '') AS _odbc_fwd_count_wrapped
DEBUG:  Successful result: Executing ODBC query
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.

If I don't set the debug level, the result is exactly the same as in my first post. Any ideas?

**Comment by pramsey** [comment]
Hm, actually no matter what version I use, the driver never pushes a qual to the remote:
```
DEBUG:  Executing query: SELECT "pk","name","birth","age","height" FROM "public"."local_table"
```
I guess the final test would be pulling pre-change copy and seeing if it works.

**Comment by pramsey** [comment]
Pulling the old version c79ab335d2de18e9ea28ce4390eabda7fcaf1a8f, I'm still seeing an unqualified query:
```
DEBUG:  Executing query: SELECT "pk","name","birth","age","height" FROM "public"."local_table"
```

**Comment by pramsey** [comment]
The crash in debug @jequiroga experienced has been fixed with 3bf83abc83a3b901904e8588ef853e8cfaf983f1 (whoops)

**Comment by jequiroga** [comment]
Hi guys, I tried with ogr_fdw, but I forgot how to make the creation of the foreign table and I haven't found the way to do it using a view or through a query. However, I tested tds_fdw and it worked very nicely, these are the results.

"Foreign Scan on ra.clients_csgor_view  (cost=25.00..17918.00 rows=17893 width=459) (actual time=259.358..505.141 rows=1 loops=1)"
"  Output: client_id, client_sin, first_name, last_name, last_app_id, app_program_year_id, app_program_year, app_award_limit, app_status_id, current_program_year_id, current_program_year, pd_established"
"  Filter: (clients_csgor_view.client_sin = '721362009'::bpchar)"
"  Rows Removed by Filter: 17892"
"Planning time: 339.110 ms"
"Execution time: 517.837 ms"

I forgot to mention that in the development environment (Windows 10) I actually used odbc_fdw because the scripts should not be different (I used before ogr_fdw, but due to the decision of maintain the minimum dependencies for the production and easier definition I selected odbc_fdw), so maybe there is something wrong with unixODBC or the way odbc_fdw interact with unixODBC which is the only piece that is not common and tds_fdw is not using it AFAIK.

So let me know what I should do (which version to re-compile) to do a final test.

Thanks for your time and attention to this issue it is much appreciated.

**Comment by pramsey** [comment]
Could you get the `DEBUG` output from both a "working" and a "slow" install of `odbc_fdw`?

**Comment by pramsey** [comment]
One thing I note in a test query is that if I make my test table 20K records in size, doing a single-record query takes almost 3s. That's the cost of pulling the whole table contents from remote to local and doing the filter locally, which is happening because the qual is not being shipped to the remote. That said, even using PgSQL 9.5 and a copy of the FDW that pre-dates our changes, I still wasn't seeing the qual being shipped. I look forward to seeing those logs to see if the "good" version you have running is in fact shipping the qual. Hard to see how it could log sub-second times without shipping the qual.

**Comment by pramsey** [comment]
Another note; putting the 3s query in the profiler, all the time is used up in the unixODBC drivers themselves, not the FDW. So all I have left is wondering if somehow your older code is in fact shipping the qual to the remote while ours is not.
![screenshot_518](https://user-images.githubusercontent.com/113883/36816955-5a208948-1c94-11e8-961a-1fa4b29e39fc.png)

**Comment by jequiroga** [comment]
Sorry I tried to recompile with the updated master, but I could not and cannot afford more time in this issue. However I suspected about the findings pramsey did, that somehow the whole data is transferred and then filtered (which explains why the other query took 3 secs to get one record). For now I have to use tds_fdw, but it seemed to me not very stable though. Once I have more time to see this issue I'll be back. 

Thanks again and I think you are already in good direction. Hoping I can help again soon enough or reproduce the issue in my spare time in a virtual environment.
